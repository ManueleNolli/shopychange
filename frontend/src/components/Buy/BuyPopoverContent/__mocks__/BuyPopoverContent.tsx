import { divWithoutChildrenMock } from '../../../../../mocksUtils'

const BuyPopoverContentMock = () => divWithoutChildrenMock('BuyPopoverContent')
export default BuyPopoverContentMock
