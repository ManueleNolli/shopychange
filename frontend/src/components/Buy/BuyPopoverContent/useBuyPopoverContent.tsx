import { useToast } from '@chakra-ui/react'
import { useUserContext } from '../../../context/userContext'
import useLoading from '../../../hooks/useLoading'

import {
  handlerMutation,
  AddNFTCollectionsMutation,
} from '../../../services/BackendService'

import { waitTransaction, buy } from '../../../services/BlockchainService'

import { BuyPopoverProps } from '../../../types/props/BuyPopoverContentProps'

export default function useBuyPopoverContent({
  nft,
  price,
  onClose,
  onSuccess,
}: BuyPopoverProps) {
  const toast = useToast()
  const { userAddress, blockchainNetworkId } = useUserContext()
  const { isLoading, setIsLoading } = useLoading()

  const buyToken = async () => {
    if (!userAddress || !blockchainNetworkId) return

    setIsLoading(true)
    return (
      buy(blockchainNetworkId, nft.contractAddress, nft.tokenId, price)
        .then(waitTransaction)
        .then(() => {
          onSuccess()
          onClose()
          setIsLoading(false)
        })
        .then(() => {
          handlerMutation(
            AddNFTCollectionsMutation([nft.contractAddress], userAddress)
          )
        })
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        .catch((err: any) => {
          // First phrase of the error message is enough
          const errorMessage = err.message.split('.')[0]
          toast({
            variant: 'solid',
            title: 'Transaction failed',
            description: errorMessage,
            status: 'warning',
            duration: 9000,
            isClosable: true,
          })
          setIsLoading(false)
        })
    )
  }

  return { isLoading, buyToken }
}
