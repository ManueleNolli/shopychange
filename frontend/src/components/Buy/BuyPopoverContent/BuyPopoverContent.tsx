import React from 'react'

import {
  Flex,
  Center,
  ButtonGroup,
  Heading,
  Text,
  Spacer,
  Container,
} from '@chakra-ui/react'

import LoadingButton from '../../LoadingButton/LoadingButton'
import ShadowButton from '../../ShadowButton/ShadowButton'
import PriceDistrubution from '../../PriceDistribition/PriceDistribution'

import { colors } from '../../../styles/Colors'

import { BuyPopoverProps } from '../../../types/props/BuyPopoverContentProps'
import useBuyPopoverContent from './useBuyPopoverContent'

export default function CancelSalePopoverContent({
  nft,
  price,
  onClose,
  onSuccess,
}: BuyPopoverProps) {
  const { isLoading, buyToken } = useBuyPopoverContent({
    nft,
    price,
    onClose,
    onSuccess,
  })

  return (
    <Flex direction="column" flex="1" p="1vh">
      <Spacer />

      <Container flex="1" centerContent>
        <Heading size="lg" textAlign={'center'}>
          Are you sure you want to buy <Text as="i">{nft.name}</Text>?
        </Heading>
      </Container>

      <Spacer />
      <PriceDistrubution
        price={price}
        isOwner={false}
        address={nft.contractAddress}
        tokenId={nft.tokenId}
      />
      <Spacer />

      <Center flex="1" mt="5">
        <ButtonGroup spacing="10">
          <LoadingButton isLoading={isLoading} data-testid="loading-button">
            <ShadowButton
              data-testid="buy-button"
              label={`Buy for ${price} ETH`}
              bgColor={`${colors.buyChakraColor}.500`}
              textColor={'white'}
              shadowColor={colors.buyShadowColor}
              hoverColor={`${colors.buyChakraColor}.600`}
              focusColor={`${colors.buyChakraColor}.600`}
              onClick={buyToken}
            />
          </LoadingButton>
          <ShadowButton
            label={'Cancel'}
            bgColor={`${colors.cancelChakraColor}.500`}
            textColor={'white'}
            shadowColor={colors.cancelShadowColor}
            hoverColor={`${colors.cancelChakraColor}.600`}
            focusColor={`${colors.cancelChakraColor}.600`}
            onClick={onClose}
          />
        </ButtonGroup>
      </Center>
      <Spacer />
    </Flex>
  )
}
