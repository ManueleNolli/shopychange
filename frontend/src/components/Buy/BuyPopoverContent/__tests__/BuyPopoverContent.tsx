import React from 'react'
import { render } from '@testing-library/react'

import BuyPopoverContent from '../BuyPopoverContent'
import useBuyPopoverContent from '../useBuyPopoverContent'

const mockProps = {
  nft: {
    contractAddress: '0x0000000000',
    tokenId: 0,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q',
    name: 'Test NFT',
  },
  onClose: jest.fn(),
  price: 0.4,
}

jest.mock('../useBuyPopoverContent', () => jest.fn())
jest.mock('../../../PriceDistribition/PriceDistribution')

const onUpdate = jest.fn()

describe('SellPopoverContent', () => {
  test('renders correctly on loading', () => {
    ;(useBuyPopoverContent as jest.Mock).mockReturnValueOnce({
      isLoading: true,
      buyToken: jest.fn(),
    })

    const tree = render(
      <BuyPopoverContent
        nft={mockProps.nft}
        onClose={mockProps.onClose}
        price={mockProps.price}
        onSuccess={onUpdate}
      />
    )

    expect(tree.container).toMatchSnapshot()
  })
  test('renders correctly', () => {
    ;(useBuyPopoverContent as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      buyToken: jest.fn(),
    })

    const tree = render(
      <BuyPopoverContent
        nft={mockProps.nft}
        onClose={mockProps.onClose}
        price={mockProps.price}
        onSuccess={onUpdate}
      />
    )

    expect(tree.container).toMatchSnapshot()
  })
})
