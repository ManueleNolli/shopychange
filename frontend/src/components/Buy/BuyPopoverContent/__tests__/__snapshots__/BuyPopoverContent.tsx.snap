// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellPopoverContent renders correctly 1`] = `
<div>
  <div
    class="css-1wjy6p3"
  >
    <div
      class="css-17xejub"
    />
    <div
      class="chakra-container css-11uhf22"
    >
      <h2
        class="chakra-heading css-xi606m"
      >
        Are you sure you want to buy 
        <i
          class="chakra-text css-0"
        >
          Test NFT
        </i>
        ?
      </h2>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      data-testid="PriceDistribution"
    />
    <div
      class="css-17xejub"
    />
    <div
      class="css-1ozwi6w"
    >
      <div
        class="chakra-button__group css-s2zvk0"
        data-orientation="horizontal"
        role="group"
      >
        <button
          class="chakra-button css-j18z53"
          data-testid="onClick-button"
          style="white-space: initial;"
          type="button"
        >
          Buy for 0.4 ETH
        </button>
        <button
          class="chakra-button css-lnz0i6"
          data-testid="onClick-button"
          style="white-space: initial;"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
    <div
      class="css-17xejub"
    />
  </div>
</div>
`;

exports[`SellPopoverContent renders correctly on loading 1`] = `
<div>
  <div
    class="css-1wjy6p3"
  >
    <div
      class="css-17xejub"
    />
    <div
      class="chakra-container css-11uhf22"
    >
      <h2
        class="chakra-heading css-xi606m"
      >
        Are you sure you want to buy 
        <i
          class="chakra-text css-0"
        >
          Test NFT
        </i>
        ?
      </h2>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      data-testid="PriceDistribution"
    />
    <div
      class="css-17xejub"
    />
    <div
      class="css-1ozwi6w"
    >
      <div
        class="chakra-button__group css-s2zvk0"
        data-orientation="horizontal"
        role="group"
      >
        <button
          class="chakra-button css-1b928yd"
          data-loading=""
          data-testid="loading-button"
          disabled=""
          type="button"
        >
          <div
            class="chakra-button__spinner chakra-button__spinner--start css-1vsp78f"
          >
            <span
              style="display: inherit;"
            >
              <span
                style="display: inline-block; background-color: white; width: 8px; height: 8px; margin: 2px; border-radius: 100%; animation: react-spinners-BeatLoader-beat 0.7s 0s infinite linear; animation-fill-mode: both;"
              />
              <span
                style="display: inline-block; background-color: white; width: 8px; height: 8px; margin: 2px; border-radius: 100%; animation: react-spinners-BeatLoader-beat 0.7s 0.35s infinite linear; animation-fill-mode: both;"
              />
              <span
                style="display: inline-block; background-color: white; width: 8px; height: 8px; margin: 2px; border-radius: 100%; animation: react-spinners-BeatLoader-beat 0.7s 0s infinite linear; animation-fill-mode: both;"
              />
            </span>
          </div>
          <span
            class="css-q7lffx"
          />
        </button>
        <button
          class="chakra-button css-lnz0i6"
          data-testid="onClick-button"
          style="white-space: initial;"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
    <div
      class="css-17xejub"
    />
  </div>
</div>
`;
